src/language_concepts/generics.md

```diff
@@ -0,0 +1,34 @@
+# Generics
+
+Generics allow you to use the same functions with multiple different concrete data types. You can read more about the concept of generics in the Rust documentation [here](https://doc.rust-lang.org/book/ch10-01-syntax.html).
+
+For example, you may need functions that add `u8` and `u32` integers. Instead of having to write two functions that do the same thing, you can write one function that works for both using generics.
+
+```rust noplaypen
+fn add<T>(a : T, b : T) -> T {
+  return a + b
+}
+```
+
+This concept also works with implementations of structs. For example
+
+```rust noplaypen
+struct BigInt<N> {
+    limbs: [u32; N],
+}
+
+impl<N> BigInt<N> {
+    // `N` is in scope of all methods in the impl
+    fn first(first: BigInt<N>, second: BigInt<N>) -> Self {
+        constrain first.limbs != second.limbs;
+        first
+    }
+
+    fn second(first: BigInt<N>, second: Self) -> Self {
+        constrain first.limbs != second.limbs;
+        second
+    }
+}
+```
+
```

Comment:
As the last example we should mention an example of how to write a function generic over all types T, where T must have some function available. For example, how can we write a generic equality function for arrays of any type? By passing in the required function on T as a parameter:
```rs
fn array_eq<T, N>(array1: [T; N], array2: [T; N], elem_eq: fn(T, T) -> bool) -> bool {
    if array1.len() != array2.len() {
        false
    } else {
        let mut result = true;
        for i in 0 .. array1.len() {
            result &= elem_eq(array1[i], array2[i]);
        }
        result
    }
}

fn main() {
    constrain array_eq([1, 2, 3], [1, 2, 3], |a, b| a == b);

    // We can use array_eq even for arrays of structs, as long as we have
    // an equality function for these structs we can pass in
    let array = [MyStruct::new(), MyStruct::new()];
    constrain array_eq(array, array, MyStruct::eq);
}